File: packages/core/src/markdown/slug.ts
/**
 * Title → filename slug derivation (Plan 17). The slug is a *projection* of
 * the title — the only author of regular-note filenames in-app — so its output
 * must be safe on every filesystem a graph can sync to. Lowercase-only output
 * is load-bearing: it makes APFS/NTFS case-insensitivity and git
 * case-sensitivity agree by construction. Non-Latin scripts pass through
 * untransliterated; a CJK title keeps its characters.
 *
 * The rules are frozen by the golden corpus in `slug.test.ts`: a silent change
 * here would re-slug every title differently — a rename storm across graphs.
 */

/**
 * Windows reserved device names (case-insensitive, extension-less). A file
 * named `con.md` is uncreatable or hazardous on Windows, so these slugs get a
 * `-note` suffix.
 */
const WINDOWS_RESERVED = new Set([
  'con',
  'prn',
  'aux',
  'nul',
  'com1',
  'com2',
  'com3',
  'com4',
  'com5',
  'com6',
  'com7',
  'com8',
  'com9',
  'lpt1',
  'lpt2',
  'lpt3',
  'lpt4',
  'lpt5',
  'lpt6',
  'lpt7',
  'lpt8',
  'lpt9',
])

/**
 * Maximum slug length in code points. Titles can be sentences; filenames have
 * byte budgets (255 on APFS/ext4/NTFS). 80 code points caps the worst case
 * (every point 3 UTF-8 bytes) comfortably under the limit with the `notes/`
 * prefix, `.md` suffix, and a collision suffix to spare.
 */
const MAX_SLUG_CHARS = 80

/** Anything that isn't a letter, number, or separator is dropped outright. */
const STRIP_RE = /[^\p{L}\p{N}\s_-]+/gu
/** Separator runs (whitespace, `_`, `-`) collapse to a single `-`. */
const SEPARATOR_RE = /[\s_-]+/gu
const EDGE_DASHES_RE = /^-+|-+$/g

/**
 * Derive the filename slug for a note title: NFC-normalize, lowercase
 * (Unicode-aware), drop everything but letters/numbers/separators, collapse
 * separator runs to single `-`, trim edge dashes, cap at
 * {@link MAX_SLUG_CHARS} code points on a character boundary. Never empty
 * (`untitled`), never a Windows reserved device name. Idempotent: a slug
 * slugs to itself.
 */
export function slugForTitle(title: string): string {
  const folded = title.normalize('NFC').toLowerCase()
  const dashed = folded
    .replace(STRIP_RE, '')
    .replace(SEPARATOR_RE, '-')
    .replace(EDGE_DASHES_RE, '')
  // Cap on code points (never split a surrogate pair), then re-trim: the cut
  // can land right after a dash.
  const capped = [...dashed].slice(0, MAX_SLUG_CHARS).join('').replace(EDGE_DASHES_RE, '')
  if (capped === '') {
    return 'untitled'
  }
  if (WINDOWS_RESERVED.has(capped)) {
    return `${capped}-note`
  }
  return capped
}

File: apps/cli/src/slug.rs
//! Title-derived note filenames, mirroring `packages/core/src/markdown/slug.ts`.

use unicode_normalization::UnicodeNormalization;

const MAX_SLUG_CHARS: usize = 60;

fn is_windows_reserved(value: &str) -> bool {
    matches!(
        value,
        "con"
            | "prn"
            | "aux"
            | "nul"
            | "com1"
            | "com2"
            | "com3"
            | "com4"
            | "com5"
            | "com6"
            | "com7"
            | "com8"
            | "com9"
            | "lpt1"
            | "lpt2"
            | "lpt3"
            | "lpt4"
            | "lpt5"
            | "lpt6"
            | "lpt7"
            | "lpt8"
            | "lpt9"
    )
}

/// Derive a portable, lowercase filename slug from a note title.
pub fn slug_for_title(title: &str) -> String {
    let mut slug = String::new();
    let mut separator = false;
    for character in title.nfc().flat_map(char::to_lowercase) {
        if character.is_alphanumeric() {
            if separator && !slug.is_empty() {
                slug.push('-');
            }
            separator = false;
            slug.push(character);
        } else if character.is_whitespace() || matches!(character, '-' | '_') {
            separator = true;
        }
    }
    slug = slug.chars().take(MAX_SLUG_CHARS).collect();
    while slug.ends_with('-') {
        slug.pop();
    }
    if slug.is_empty() {
        return "untitled".to_string();
    }
    if is_windows_reserved(&slug) {
        return format!("{slug}-note");
    }
    slug
}

#[cfg(test)]
mod tests {
    use super::slug_for_title;

    #[test]
    fn matches_the_core_slug_examples() {
        assert_eq!(slug_for_title("Meeting Notes"), "meeting-notes");
        assert_eq!(slug_for_title("Don't Panic!"), "dont-panic");
        assert_eq!(slug_for_title("Q3 / Q4 Review"), "q3-q4-review");
        assert_eq!(slug_for_title("path\\with\\slashes"), "pathwithslashes");
        assert_eq!(slug_for_title("日本語ノート"), "日本語ノート");
        assert_eq!(slug_for_title("🎉🎉🎉"), "untitled");
        assert_eq!(slug_for_title("CON"), "con-note");
        assert_eq!(slug_for_title(&"a".repeat(80)), "a".repeat(60));
    }
}
